local Foundation = script.Parent.Parent.Parent
local PackageManager = require(Foundation.PackageManager)

local RbxDesignFoundations = PackageManager.getPackage("RBXDesignFoundations")

local Device = require(Foundation.Enums.Device)
local Theme = require(Foundation.Enums.Theme)

type Theme = Theme.Theme
type Device = Device.Device

local FoundationDisableTokenScaling = false

local function getPlatformScale(device: Device, scaleFactor: number?)
	if FoundationDisableTokenScaling then
		return 1
	end

	scaleFactor = if scaleFactor ~= nil then scaleFactor else 1
	scaleFactor = math.clamp(scaleFactor :: number, 0, math.huge)
	-- Platform scale will be from engine API as soon as it's ready.
	-- For now scale values are hard-coded, and only console uses 1.5
	-- differently according to design specs.
	local baseScale = if device == Device.Console then 1.5 else 1

	return baseScale * scaleFactor :: number
end

local function getTokens(theme: Theme, deviceInput: Device?, scaleFactor: number?)
	local generators = RbxDesignFoundations.Tokens
	local device: Device = deviceInput or Device.Desktop
	local scale = getPlatformScale(device, scaleFactor)
	local themeTokens: typeof(generators.Dark) = if theme == Theme.Dark then generators.Dark else generators.Light

	local tokens = themeTokens(scale)

	local filteredTokens = {
		Color = tokens.Color,
		Config = tokens.Config,
		DarkMode = tokens.DarkMode,
		Ease = tokens.Ease,
		LightMode = tokens.LightMode,
		Inverse = tokens.Inverse,
		Semantic = tokens.Semantic, -- TODO(tokens): Remove once icon sizes are moved
		Size = tokens.Size,
		Padding = tokens.Padding,
		Margin = tokens.Margin,
		Gap = tokens.Gap,
		Gutter = tokens.Gutter,
		Radius = tokens.Radius,
		Stroke = tokens.Stroke,
		LineHeight = tokens.LineHeight,
		FontSize = tokens.FontSize,
		FontWeight = tokens.FontWeight,
		Time = tokens.Time,
		Typography = tokens.Typography,
	}

	return filteredTokens
end

local defaultTokens = getTokens(Theme.Dark, Device.Desktop)
export type Tokens = typeof(defaultTokens)

return {
	getTokens = getTokens,
	defaultTokens = defaultTokens,
}
